cmd/pebble/cmd_exec.go

```diff
@@ -0,0 +1,269 @@
+	if ok {
+		env["TERM"] = term
+	}
+	for _, kv := range cmd.Env {
+		parts := strings.SplitN(kv, "=", 2)
+		key := parts[0]
+		value := ""
+		if len(parts) == 2 {
+			value = parts[1]
+		}
+		env[key] = value
+	}
+
+	// Send UID if it looks like an integer, otherwise send user name.
+	var user string
+	var userID *int
+	uid, err := strconv.Atoi(cmd.User)
+	if err != nil {
+		user = cmd.User
+	} else {
+		userID = &uid
+	}
+
+	// Send GID if it looks like an integer, otherwise send group name.
+	var group string
+	var groupID *int
+	gid, err := strconv.Atoi(cmd.Group)
+	if err != nil {
+		group = cmd.Group
+	} else {
+		groupID = &gid
+	}
+
+	// Determine interaction mode
+	stdinTerminal := ptyutil.IsTerminal(unix.Stdin)
+	stdoutTerminal := ptyutil.IsTerminal(unix.Stdout)
+	var terminal bool
+	if cmd.Terminal {
+		terminal = true
+	} else if cmd.NoTerminal {
+		terminal = false
+	} else {
+		terminal = stdinTerminal && stdoutTerminal
+	}
+
+	// Record terminal state
+	if terminal && stdinTerminal {
+		oldState, err := ptyutil.MakeRaw(unix.Stdin)
+		if err != nil {
+			return err
+		}
+		defer ptyutil.Restore(unix.Stdin, oldState)
+	}
+
+	// Grab current terminal dimensions
+	var width, height int
+	if stdoutTerminal {
+		width, height, err = ptyutil.GetSize(unix.Stdout)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Run the command
+	opts := &client.ExecOptions{
+		Command:     command,
+		Environment: env,
+		WorkingDir:  cmd.WorkingDir,
+		Timeout:     cmd.Timeout,
+		User:        user,
+		UserID:      userID,
+		Group:       group,
+		GroupID:     groupID,
+		Terminal:    terminal,
+		Stderr:      !terminal,
+		Width:       width,
+		Height:      height,
+	}
+	additionalArgs := &client.ExecAdditionalArgs{
+		Stdin:  os.Stdin,
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+		Control: func(conn *websocket.Conn) {
+			execControlHandler(conn, terminal)
+		},
+		DataDone: make(chan bool),
+	}
+	changeID, err := cmd.client.Exec(opts, additionalArgs)
+	if err != nil {
+		return err
+	}
+
+	// Wait till the command (change) is finished
+	waitOpts := &client.WaitChangeOptions{}
+	if cmd.Timeout != 0 {
+		// A little more than the command timeout to ensure that happens first
+		waitOpts.Timeout = cmd.Timeout + time.Second
+	}
+	change, err := cmd.client.WaitChange(changeID, waitOpts)
+	if err != nil {
+		return err
+	}
+	if !change.Ready {
+		// This case shouldn't happen (because the change should either be ready,
+		// or if a timeout was specified it will apply at the Exec level and the
+		// change will be Ready but in an error state), but handle just in case.
+		return errors.New("command unexpectedly not finished after waiting for change")
+	}
+	if change.Err != "" {
+		return errors.New(change.Err)
+	}
+	var returnCode int
+	err = change.Get("return", &returnCode)
+	if err != nil {
+		return err
+	}
+	if returnCode != 0 {
+		logger.Debugf("Process exited with return code %d", returnCode)
+		panic(&exitStatus{returnCode})
```

Comment:
Yeah, this is a weird one -- but this `exitStatus` panic value is actually recovered by `main()` to return a non-zero exit status. I guess they did it that way to avoid an inline `os.Exit()` to make things more testable? I dunno. In any case, I may still remove it when I write the tests if I do that another way.